File: problem/ai-battle-series/02-format-matters/src/main.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n = 0;
    int q = 0;
    if (!(cin >> n >> q)) {
        return 0;
    }

    vector<double> prefix_score(n + 1, 0.0);
    vector<double> prefix_weight(n + 1, 0.0);

    for (int i = 1; i <= n; i++) {
        double score = 0.0;
        double weight = 0.0;
        cin >> score >> weight;
        prefix_score[i] = prefix_score[i - 1] + score * weight;
        prefix_weight[i] = prefix_weight[i - 1] + weight;
    }

    cout << fixed << setprecision(6);
    for (int i = 0; i < q; i++) {
        int l = 0;
        int r = 0;
        cin >> l >> r;
        double sum_score = prefix_score[r] - prefix_score[l - 1];
        double sum_weight = prefix_weight[r] - prefix_weight[l - 1];
        double avg = sum_weight != 0.0 ? sum_score / sum_weight : 0.0;
        cout << avg << "
";
    }
    return 0;
}

File: problem/ai-battle-series/02-format-matters/src/main.py
import sys


def main():
    data = sys.stdin.read().strip().split()
    if not data:
        return
    it = iter(data)
    n = int(next(it))
    q = int(next(it))

    prefix_score = [0.0] * (n + 1)
    prefix_weight = [0.0] * (n + 1)

    for i in range(1, n + 1):
        score = float(next(it))
        weight = float(next(it))
        prefix_score[i] = prefix_score[i - 1] + score * weight
        prefix_weight[i] = prefix_weight[i - 1] + weight

    out_lines = []
    for _ in range(q):
        l = int(next(it))
        r = int(next(it))
        sum_score = prefix_score[r] - prefix_score[l - 1]
        sum_weight = prefix_weight[r] - prefix_weight[l - 1]
        avg = sum_score / sum_weight if sum_weight else 0.0
        out_lines.append(f"{avg:.6f}")
    sys.stdout.write("\n".join(out_lines) + "\n")


if __name__ == "__main__":
    main()
